runtime/sema/check_composite_declaration.go

```diff
@@ -61,6 +61,72 @@ func (checker *Checker) checkAttachmentBaseType(attachmentType *CompositeType) {
 	})
 }
 
+func (checker *Checker) checkAttachmentMembersAccess(attachmentType *CompositeType) {
+
+	// all the access modifiers for attachment members must be elements of the
+	// codomain of the attachment's entitlement map. This is because the codomain
+	// of the attachment's declared map specifies all the entitlements one can possibly
+	// have to that attachment, since the only way to obtain an attachment reference
+	// is to access it off of a base (and hence through the map).
+	// ---------------------------------------------------
+	// entitlement map M {
+	//     E -> F
+	//     X -> Y
+	//     U -> V
+	// }
+	//
+	// access(M) attachment A for R {
+	//	  access(F) fun foo() {}
+	//	  access(Y | F) fun bar() {}
+	//	  access(V & Y) fun baz() {}
+	//
+	//    access(V | Q) fun qux() {}
+	// }
+	// ---------------------------------------------------
+	//
+	// in this example, the only entitlements one can ever obtain to an &A reference are
+	// `F`, `Y` and `V`, and as such these are the only entitlements that may be used
+	// in `A`'s definition.  Thus the definitions of `foo`, `bar`, and `baz` are valid,
+	// while the definition of `qux` is not.
+	var attachmentAccess Access = UnauthorizedAccess
+	if attachmentType.attachmentEntitlementAccess != nil {
+		attachmentAccess = *attachmentType.attachmentEntitlementAccess
+	}
+
+	attachmentType.Members.Foreach(func(_ string, member *Member) {
+		switch attachmentAccess := attachmentAccess.(type) {
+		case EntitlementMapAccess:
+			codomain := attachmentAccess.Codomain()
+			switch memberAccess := member.Access.(type) {
+			case EntitlementSetAccess:
+				memberAccess.Entitlements.Foreach(func(entitlement *EntitlementType, _ struct{}) {
+					if !codomain.Entitlements.Contains(entitlement) {
+						if !member.Access.Equal(UnauthorizedAccess) {
+							checker.report(&InvalidAttachmentEntitlementError{
```

Comment:
inversing these two conditions may help reduce the number of nested levels.

Reply:
Looks like the second condition here is actually unnecessary, so I removed it.